### PR TITLE
Chapter9

### PR DESCRIPTION
**Pull request by NurAnisAqilah**
ready

### CODE REVIEW

**Comment by NurAnisAqilah** [comment]
ready

File: Gemfile [hunk 1]
```diff
@@ -7,6 +7,8 @@ gem "rails", "~> 6.0.3", ">= 6.0.3.2"
 
 gem "mysql2"
 
+gem "sqlite3"
```
Comment (ngocvt-0484) [inline Gemfile]: remove it because you are using mysql2

**Comment by NurAnisAqilah** [comment]
ready

File: Gemfile.lock [hunk 2]
```diff
@@ -197,6 +197,7 @@ GEM
       actionpack (>= 4.0)
       activesupport (>= 4.0)
       sprockets (>= 3.0.0)
+    sqlite3 (1.4.2)
```
Comment (ngocvt-0484) [inline Gemfile.lock]: please re bundle again
bundle install

**Comment by NurAnisAqilah** [comment]
ready

File: app/models/user.rb [hunk 3]
```diff
@@ -1,19 +1,46 @@
 class User < ApplicationRecord
-  validates :name, presence: true,
-    length: {maximum: Settings.user.name.max_length}
-  
-  validates :email, presence: true,
-    length: {maximum: Settings.user.email.max_length},
-    format: {with: Settings.user.email.regex},
+  attr_accessor :remember_token
+  validates :name, presence: true,
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
  attr_accessor :remember_token
  
  validates :name, presence: true, 
```

File: app/models/user.rb [hunk 4]
```diff
@@ -1,19 +1,46 @@
...
+ 
+  def remember
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest remember_token
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    Bcrypt::Password.new(remember_digest).is_password? remember_token
+  end
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
    return false if remember_digest.nil?
    
    Bcrypt::Password.new(remember_digest).is_password? remember_token
  end
```

**Comment by NurAnisAqilah** [comment]
ready

File: app/models/user.rb [hunk 5]
```diff
@@ -1,19 +1,48 @@
...
-    length: {minimum: Settings.user.password.min_length}
+  validates :password, presence: true, 
+    length: { minimum: Settings.validations.password.min_length }
   
-  has_secure_password
+  has_secure_password 
   
   before_save :downcase_email
   
+  class << User
```
Comment (manhnt-0960) [inline app/models/user.rb]:
```suggestion
  class << self
```

File: app/models/user.rb [hunk 6]
```diff
@@ -1,19 +1,48 @@
...
+    end
+  end
+ 
+  def remember
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest remember_token
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
```
Comment (manhnt-0960) [inline app/models/user.rb]:
```suggestion
    return false unless remember_digest
```

**Review by manhnt-0960 — COMMENTED** [review]
please fix then merge it

**Pull request merged**